File: packages/core/src/entity/playerEntity.ts
import { z } from 'zod';

import { RaceTypeSchema } from '../domain/model/valueObject/raceType';

/**
 * 選手情報を表すエンティティのzod型定義
 *
 * 競馬・競輪などの選手情報を保持します。
 * バリデーションルール：
 * - raceTypeは有効なレース種別である必要があります。
 * - playerNo、playerNameは空でない文字列である必要があります。
 * - priorityは0以上の整数である必要があります。
 */
export const PlayerEntitySchema = z.object({
    /** レース種別（JRA/NAR/KEIRINなど） */
    raceType: RaceTypeSchema,
    /** プレイヤー番号 */
    playerNo: z.string().min(1, 'playerNo must not be empty'),
    /** プレイヤー名 */
    playerName: z.string().min(1, 'playerName must not be empty'),
    /** 優先度 */
    priority: z
        .number()
        .int('priority must be an integer')
        .min(0, 'priority must be non-negative'),
    /** 期別（選手養成所の卒業期。KEIRINのみ、player_keirinから補う。省略可） */
    term: z.number().int().positive('term must be positive').optional(),
    /** 所属（KEIRIN=府県、AUTORACE=拠点/LG。player_keirin/player_autoraceから補う。省略可） */
    branch: z.string().min(1, 'branch must not be empty').optional(),
});

/**
 * PlayerEntityの型定義
 */
export type PlayerEntity = z.infer<typeof PlayerEntitySchema>;

/**
 * PlayerEntityのバリデーション関数
 * @param entity - バリデーション対象のPlayerEntityオブジェクト
 * @returns バリデーション済みのPlayerEntityオブジェクト
 * @throws バリデーションエラーが発生した場合はzodのエラーをスローします。
 */
export const validatePlayerEntity = (entity: unknown): PlayerEntity => {
    return PlayerEntitySchema.parse(entity);
};

File: packages/front/lib/domain/entities/player_entity.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'race_type.dart';

part 'player_entity.freezed.dart';

/// 注目選手として登録するときの `priority` 値（KPLAYER-07）。
/// バックエンド（player_watch.priority）の「0=注目しない/10=注目する」の
/// 二値運用にそのまま揃える。
const int kWatchedPlayerPriority = 10;

/// 注目選手機能の対象レース種別（KPLAYER-07 → AUTORACE拡張）。
///
/// 選手データ基盤（player_keirin/player_autorace）が整っているKEIRIN・
/// AUTORACEのみが対象。BOATRACE等は選手データの取得経路自体が無いため対象外。
const watchedPlayerRaceTypes = [RaceType.keirin, RaceType.autorace];

@freezed
abstract class PlayerEntity with _$PlayerEntity {
  const factory PlayerEntity({
    // jra, nar, keirin, overseas, autorace, boatrace
    required String raceType,
    // 選手コード（先頭ゼロを保持するため文字列）
    required String playerNo,
    required String playerName,
    // 0 = 注目しない / 10 = 注目する（KPLAYER-07）。
    // バックエンド（player_watch.priority）の二値運用にそのまま揃える。
    required int priority,
    // 期別（選手養成所の卒業期。KEIRINのみ、省略され得る）
    int? term,
    // 所属（KEIRINは府県、AUTORACEは拠点/LG。省略され得る）
    String? branch,
  }) = _PlayerEntity;
}
